File: Bheem Wants Ladoos [20 - April - 2023]/Prefix Suffix String[ 21 - April - 2023]/solution.py
#   >> PYTHON CODE <<

class Solution:
    class TrieNode:
        def __init__(self):
            self.isEndOfWord=False
            self.children=[None]*26
        
        def insert(self,s1,root):
            cur=root
            for i in range(0,len(s1)):
                index=ord(s1[i])-ord('a')
                if(cur.children[index] is None):
                    cur.children[index]=Solution.TrieNode()
                
                cur=cur.children[index]
            
            cur.isEndOfWord=True
        
        def search(self,s1,root):
            cur=root
            for i in range(0,len(s1)):
                index=ord(s1[i])-ord('a')
                if(cur.children[index] is None):
                    return False
                
                cur=cur.children[index]
            
            return True
            
    def prefixSuffixString(self, s1, s2) -> int:
        
        root=self.TrieNode()
        
        for val in s1:
            root.insert(val[::-1],root)
            root.insert(val,root)
        
        ans=0;
        
        for val in s2:
            if(root.search(val,root) or root.search(val[::-1],root)):
                ans+=1
        
        return ans

File: Bheem Wants Ladoos [20 - April - 2023]/Prefix Suffix String[ 21 - April - 2023]/solution.java
//   >> JAVA CODE <<

class Solution
{
    class TrieNode{
        boolean isEndOfWord;
        TrieNode[] children;
        
        TrieNode()
        {
            this.isEndOfWord=true;
            this.children=new TrieNode[26];
        }
    }   
    
    public void insert(String s,TrieNode root)
    {
        TrieNode cur=root;
        
        for(int i=0;i<s.length();i++)
        {
            int index=s.charAt(i)-'a';
            
            if(cur.children[index]==null)
                cur.children[index]=new TrieNode();
                
            cur=cur.children[index];
        }
        
        cur.isEndOfWord=true;
    }
    public boolean search(String s,TrieNode root)
    {
        TrieNode cur=root;
        
        for(int i=0;i<s.length();i++)
        {
            int index=s.charAt(i)-'a';
            
            if(cur.children[index]==null)
                return false;
                
            cur=cur.children[index];
        }
        
        
        return true;
    }
    public int prefixSuffixString(String s1[],String s2[])
    {
        
        TrieNode root=new TrieNode();
        
        for(var val:s1)
        {
            StringBuffer rev=new StringBuffer(val); 
            rev.reverse();
            
            insert(val,root);
            insert(rev.toString(),root);
        }
        
        int ans=0;
        
        for(var val:s2)
        {
            StringBuffer rev=new StringBuffer(val); 
            rev.reverse();
            
            if(search(val,root) || search(rev.toString(),root))
                ans++;
        }
            
            return ans;
    }
}